File: backend-services/product-and-inventory/src/modules/category/category.service.ts
import { Injectable } from '@nestjs/common';
import { CreateCategoryDto } from './dto/requests/create-category.dto';
import { UpdateCategoryDto } from './dto/requests/update-category.dto';
import { InjectModel } from '@nestjs/mongoose';
import { Category } from './schemas/category.schema';
import mongoose, { Model } from 'mongoose';
import { RpcException } from '@nestjs/microservices';
import { ApiResponseInterface } from '../../common/dto/api-response.interface';
import { mapCategoryToResponseDto } from '../../common/mappers/category.mapper';
import { CategoryResponseDto } from './dto/responses/category-response.dto';
import { CategoryTreeResponseDto } from './dto/responses/category-tree-response.dto';
import { CategoryEvents } from '../../common/events/registry.events';

@Injectable()
export class CategoryService {
  constructor(
    @InjectModel(Category.name) private readonly categoryModel: Model<Category>,
  ) {}

  async create(
    createCategoryDto: CreateCategoryDto,
  ): Promise<ApiResponseInterface<CategoryResponseDto>> {
    const category = await this.categoryModel
      .findOne({ name: createCategoryDto.name })
      .exec();
    if (category != null) {
      throw new RpcException('Category already exists');
    }

    if (createCategoryDto.parentId != null) {
      const parentCategory = await this.categoryModel
        .findOne({ _id: createCategoryDto.parentId })
        .exec();
      if (parentCategory == null) {
        throw new RpcException('Parent category does not exist');
      }
      const newCategory = await this.categoryModel.create({
        name: createCategoryDto.name,
        parentCategory: parentCategory._id,
      });

      const responseDto = mapCategoryToResponseDto(newCategory);

      return {
        event: CategoryEvents.CREATE,
        status: 'success',
        data: responseDto,
        message: null,
      };
    }

    const newCategory = await this.categoryModel.create({
      name: createCategoryDto.name,
      parentCategory: null,
    });

    const responseDto = mapCategoryToResponseDto(newCategory);

    return {
      event: CategoryEvents.CREATE,
      status: 'success',
      message: 'Category created successfully',
      data: responseDto,
    };
  }

  async update(
    updateCategoryDto: UpdateCategoryDto,
  ): Promise<ApiResponseInterface<CategoryResponseDto>> {
    const category = await this.categoryModel
      .findOne({ _id: updateCategoryDto.id })
      .exec();
    if (category == null) {
      throw new RpcException('Category does not exist');
    }

    if (updateCategoryDto.name != null) {
      const categoryWithSameName = await this.categoryModel
        .findOne({ name: updateCategoryDto.name })
        .exec();
      if (categoryWithSameName != null) {
        throw new RpcException('Category with same name already exists');
      }
      category.name = updateCategoryDto.name;
    }

    if (updateCategoryDto.parentId != null) {
      const parentCategory = await this.categoryModel
        .findOne({ _id: updateCategoryDto.parentId })
        .exec();
      if (parentCategory == null) {
        throw new RpcException('Parent category does not exist');
      }
      category.parentCategory = new mongoose.Types.ObjectId(parentCategory._id);
    }

    await category.save();

    const responseDto = mapCategoryToResponseDto(category);
    return {
      event: CategoryEvents.UPDATE,
      status: 'success',
      message: null,
      data: responseDto,
    };
  }

  async delete(id: string): Promise<ApiResponseInterface<string>> {
    const response = await this.findAllSubcategoriesById(id);
    if (response.data !== null && response.data.length !== 0) {
      throw new RpcException('Category has subcategories');
    }
    await this.categoryModel.findByIdAndDelete(id).exec();
    return {
      event: CategoryEvents.DELETE,
      status: 'success',
      message: null,
      data: id,
    };
  }

  async findAllSubcategoriesById(
    id: string,
  ): Promise<ApiResponseInterface<CategoryResponseDto[]>> {
    const category = await this.categoryModel.findOne({ _id: id }).exec();
    if (category == null) {
      throw new RpcException('Category does not exist');
    }

    const subcategories = await this.categoryModel
      .find({ parentCategory: id })
      .exec();

    const responseDto = subcategories.map((subcategory) =>
      mapCategoryToResponseDto(subcategory),
    );
    return {
      event: CategoryEvents.FIND_SUBCATEGORIES_BY_ID,
      status: 'success',
      message: null,
      data: responseDto,
    };
  }

  async findAllParentCategories(): Promise<
    ApiResponseInterface<CategoryResponseDto[]>
  > {
    const parentCategories = await this.categoryModel
      .find({ parentCategory: null })
      .exec();

    const responseDto = parentCategories.map((category) =>
      mapCategoryToResponseDto(category),
    );
    return {
      event: CategoryEvents.FIND_ALL_PARENTS,
      status: 'success',
      message: null,
      data: responseDto,
    };
  }

  async findOneById(
    id: string,
  ): Promise<ApiResponseInterface<CategoryResponseDto>> {
    const category = await this.categoryModel.findById(id).exec();
    if (category == null) {
      throw new RpcException('Category not found');
    }

    const responseDto = mapCategoryToResponseDto(category);

    return {
      event: CategoryEvents.FIND_ONE_BY_ID,
      status: 'success',
      message: null,
      data: responseDto,
    };
  }

  async findAllCategories(): Promise<
    ApiResponseInterface<CategoryResponseDto[]>
  > {
    const categories = await this.categoryModel.find().exec();

    const responseDto = categories.map((category) =>
      mapCategoryToResponseDto(category),
    );
    return {
      event: CategoryEvents.FIND_ALL,
      status: 'success',
      message: null,
      data: responseDto,
    };
  }

  async findCategoryTree(): Promise<
    ApiResponseInterface<CategoryTreeResponseDto[]>
  > {
    // fetch all categories
    const categories = (await this.findAllCategories()).data;
    if (categories == null || categories.length === 0) {
      return {
        event: CategoryEvents.FIND_CATEGORY_TREE,
        status: 'success',
        message: null,
        data: [],
      };
    }

    const childrenMap = new Map<string, CategoryResponseDto[]>();
    categories.forEach((category) => {
      if (category.parentId) {
        if (!childrenMap.has(category.parentId)) {
          childrenMap.set(category.parentId, []);
        }
        childrenMap.get(category.parentId)!.push(category);
      }
    });

    const roots = categories
      .filter((category) => category.parentId === null)
      .map((root) => ({
        id: root.id,
        name: root.name,
        children: childrenMap.get(root.id) || [],
      }));

    return { event: CategoryEvents.FIND_CATEGORY_TREE, status: 'success', message: null, data: roots };
  }
}

File: backend-services/inventory_go/service/category_service.go
package service

import (
	"errors"
	"inventory_go/category"
	findmanyoptions "inventory_go/category/enums"
	"inventory_go/category/mappers"
	"log"
)

type CategoryServiceImpl struct {
	repository category.CategoryRepository
}

func NewCategoryService(repository category.CategoryRepository) *CategoryServiceImpl {
	return &CategoryServiceImpl{repository: repository}
}

func (service *CategoryServiceImpl) Create(category *category.Category) error {
	if category.Id() != "" {
		return errors.New("id should be empty when creating a category")
	}

	// Rule 0: Each category should have a unique name
	_category, err := service.repository.FindByName(category.Name())
	if err != nil {
		return err
	}
	if _category != nil {
		return errors.New("category already exists")
	}

	// Check if the parent category exists
	parentCategory, err := service.repository.GetById(category.ParentId())
	if err != nil {
		return err
	}
	if parentCategory == nil {
		return errors.New("parent category not found")
	}

	err = service.repository.Save(category)
	return err
}

func (service *CategoryServiceImpl) Update(id string, name string) error {
	// Rule 0: Each category should have a unique name
	_category, err := service.repository.FindByName(name)
	if err != nil {
		return err
	}
	if _category != nil {
		return errors.New("name already exists")
	}

	// Check if category exists
	_category, err = service.repository.GetById(id)
	if err != nil {
		return err
	}
	if _category == nil {
		return errors.New("category not found")
	}

	domain := mappers.ToDomain(_category)
	event, err := domain.Rename(name, true)

	// For now, log the event as nobody is listening
	log.Print(event)

	err = service.repository.Save(domain)
	return err
}

func (service *CategoryServiceImpl) Delete(id string) error {
	// Rule 1: A category cannot be deleted if it has subcategories
	subcategories, err := service.repository.FindManyByFilter(findmanyoptions.FindAllSubCategoriesById, id)
	if err != nil {
		return err
	}
	if len(subcategories) > 0 {
		return errors.New("subcategories still exists")
	}
	err = service.repository.Delete(id)
	return err
}

func (service *CategoryServiceImpl) FindOneById(id string) (*category.Category, error) {
	_category, err := service.repository.GetById(id)
	if err != nil {
		return nil, err
	}
	if _category == nil {
		return nil, errors.New("category not found")
	}
	domain := mappers.ToDomain(_category)
	return domain, err
}

func (service *CategoryServiceImpl) FindAllSubCategoriesById(id string) ([]*category.Category, error) {
	// Check if category exists
	_category, err := service.repository.GetById(id)
	if err != nil {
		return nil, err
	}
	if _category == nil {
		return nil, errors.New("category not found")
	}
	subcategories, err := service.repository.FindManyByFilter(findmanyoptions.FindAllSubCategoriesById, id)
	if err != nil {
		return nil, err
	}
	var result []*category.Category
	for _, subcategory := range subcategories {
		result = append(result, mappers.ToDomain(subcategory))
	}
	return result, err
}

func (service *CategoryServiceImpl) FindAllParentCategories() ([]*category.Category, error) {
	categories, err := service.repository.FindManyByFilter(findmanyoptions.FindAllParents, "")
	if err != nil {
		return nil, err
	}
	var result []*category.Category
	for _, _category := range categories {
		result = append(result, mappers.ToDomain(_category))
	}
	return result, err
}
func (service *CategoryServiceImpl) FindAll() ([]*category.Category, error) {
	categories, err := service.repository.FindManyByFilter(findmanyoptions.FindAll, "")
	if err != nil {
		return nil, err
	}
	var result []*category.Category
	for _, _category := range categories {
		result = append(result, mappers.ToDomain(_category))
	}
	return result, err
}

func (service *CategoryServiceImpl) FindCategoryTree() ([]*category.CategoryTreeNode, error) {
	categories, err := service.FindAll()
	if err != nil {
		return nil, err
	}
	if len(categories) == 0 || categories[0] == nil {
		return nil, errors.New("no category found")
	}

	var childrenMap map[string][]*category.Category
	for _, _category := range categories {
		if _category.ParentId() != "" {
			_, ok := childrenMap[_category.ParentId()]
			if !ok {
				childrenMap[_category.ParentId()] = []*category.Category{}
			}
			childrenMap[_category.ParentId()] = append(childrenMap[_category.ParentId()], _category)
		}
	}

	var result []*category.CategoryTreeNode
	for _, _category := range categories {
		if _category.ParentId() != "" {
			continue
		}
		children, ok := childrenMap[_category.ParentId()]
		if !ok {
			continue
		}
		result = append(result, &category.CategoryTreeNode{Id: _category.Id(), Name: _category.Name(), Children: children})
	}
	return result, err
}
